### ISSUE TITLE
Create a release GitHub workflow

### ISSUE THREAD

**Issue by krzema12** [krzema12/snakeyaml-engine-kmp#94]
E.g. after pushing a tag, a release process of building for all targets (especially Mac OS) should start.
Currently I have releasing configured on my legacy Ubuntu machine, which effectively makes this library not provide OS X targets.

**Comment by krzema12** [comment]
```
./gradlew publishAllPublicationsToSonatypeReleaseRepository
```
but closing must be done via Nexus UI manually. See how releasing is done in github-workflows-kt, the closing is done automatically there.

**Comment by OptimumCode** [comment]
Hi, @krzema12 

You can consider this Gradle plugin as a candidate here: https://github.com/vanniktech/gradle-maven-publish-plugin

I did not use it personally (I use https://github.com/gradle-nexus/publish-plugin) but many people recommend to use `gradle-maven-publish-plugin` - it hides redundant complexity under the hood.

Hope this will be helpful

**Comment by OptimumCode** [comment]
Hi, @krzema12 @aSemy.

I think I will have some time to create an initial draft for this ticket. Could you please advise if you were looking at this as well? Just wanted to check to avoid doing same thing twice)

I am going to use the maven publish plugin that I suggested before. Do you have any objections to that? I tried to apply it locally and it looks much simpler than it was before.

Regarding the workflow: I think the best option is to use `workflow_dispatch` even to trigger the release manually. What do you think about that? Alternative, the workflow can be triggered when a release/tag is published but I am not sure if this would be the best option (in my experience it is simpler to use manual trigger that will publish artifacts and create the GitHub release).

Please let me know what do you think about all the above? I would much appreciate to hear your opinion.

**Comment by aSemy** [comment]
Hey, thanks for the update @OptimumCode. I'm not looking into this, and won't have time/energy to for a while, so feel free to pick it up!

I have some thoughts on the matter though. I'll explain, because maybe it helps.

I'm personally not a fan of the publishing plugins because they implement Gradle anti-patterns. For example, [not being compatible with Configuration Cache](https://github.com/vanniktech/gradle-maven-publish-plugin/issues/259#issuecomment-845706799), and [using discouraged practices like `allprojects {}` and `afterEvaluate {}`](https://github.com/gradle-nexus/publish-plugin/blob/06a464534d1a206b7f10f49c56ef666a42d9f560/src/main/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPlugin.kt#L217-L218).

I have been interested in using GitHub Actions to help manage publishing. This action in particular looks very promising: https://github.com/nexus-actions/nexus-actions-demo#create-nexus-staging-repository-sample. 

Using these actions instead of Gradle plugins would mean releasing isn't tied to some fragile Gradle config.

**Comment by OptimumCode** [comment]
Thank you for sharing this @aSemy.

I can look at the approach using separate GitHub actions for creating and releasing staging repositories.

Also, just wanted to check how you would see the publication flow for release (considering the fact that the version change will invalidate the configuration cache if I am not mistaken):

1. commit new version without `-SNAPSHOT` suffix
2. the `build.yaml` workflow is executed (the configuration cache will be updated)
3. trigger the release workflow
    1.   create staging repo
    2.   publish artifacts (configuration cache and build cache are reused)
    3.   release/drop staging repo depending on the result of the publication
4. commit the next snapshot version

Do you have a similar picture in your mind?

If I understood the issue with the configuration cache for [gradle-maven-publish-plugin](https://github.com/vanniktech/gradle-maven-publish-plugin) correctly, if we use the plugin instead of actions the only difference will be at step **3(ii)** - the configuration cache won't be reused.

**Comment by krzema12** [comment]
Thanks @OptimumCode for taking care if it! Currently my entire OSS time budget is used by stuff related to major changes in github-workflows-kt, and also this task is not urgent (but important, I do want someone else than be to be able to release), but anyway I appreciate your work here.

I'm generally open to trying a new approach here. The workflow you described above makes sense. Regarding using a Gradle plugin or not, two thoughts:
1. I appreciate that some of them just ensapsulate the whole logic of publishing under a single task, or two tasks. On the other hand, if we control these several steps ourselves (staging repo creation, and so on), I think it's not that complex, so I'm fine with this.
2. Regarding configuration cache, my approach is more pragmatic: we release so rarely that any time savings coming from using the cache are marginal. For day-to-day building, it's of course another story.

**Comment by aSemy** [comment]
> 1. commit new version without `-SNAPSHOT` suffix
> 2. the `build.yaml` workflow is executed (the configuration cache will be updated)
> 3. trigger the release workflow
>    
>    1. create staging repo
>    2. publish artifacts (configuration cache and build cache are reused)
>    3. release/drop staging repo depending on the result of the publication
> 4. commit the next snapshot version
> 
> Do you have a similar picture in your mind?

Yeah, that sounds great!

It's not a huge problem if the configuration cache can't be re-used. This is a small project, and the cost of re-configuring is small. Regarding performance, CC will help us most with is allowing the compilation and publishing tasks to run in parallel, but even that is a relatively minor saving.

My concern with configuration cache (and other Gradle features) is less about the build performance, but more about keeping the project easier to maintain. What I really want to avoid is that some Gradle or plugin update which breaks things. I think the best way to achieve that is by avoiding reliance on discouraged Gradle practices, especially if there's an alternative.

**Comment by OptimumCode** [comment]
In this case, I will proceed with the approach that uses GitHub actions to create and release Sonatype staging repositories.

The only thing left is to decide how you want to trigger the release process and whether you want to add (depending on the way the release is triggered):

- snapshot publication to Sonatype on push to the **main** branch
- automatically generated release notes (I am personally using [mikepenz/release-changelog-builder-action](https://github.com/mikepenz/release-changelog-builder-action) for this) and publish the release to GitHub
- automatic version increment after release (automatically increments the maintenance version and commits changes)

If would prefer to generate the release notes manually, I think the best way is to trigger the workflow on [release](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#release) event.

Otherwise, I would suggest using the `workflow_dispatch` event to trigger the workflow.

**Comment by krzema12** [comment]
Releasing to Maven Central on creating a release in GitHub creates a problem: it takes 10-15 minutes for the new version to appear there, so if someone's subscribed to updates in the project, they won't be able to use the new version immediately. That's why I think we should simply trigger the workflow manually (workflow_dispatch).

Regarding automating release notes creation, we're a very low-traffic repo, so I think we're fine with crafting these manually for now, and focusing on having the core logic of publishing in place. We can always improve incrementally later.

**Comment by krzema12** [comment]
I set the secrets (although some of them may have incorrect values because how in-memory signing keys need to be exported).

Current state: https://github.com/krzema12/snakeyaml-engine-kmp/actions/runs/10488493312/job/29051106261
`> Cannot perform signing task ':signIosArm64Publication' because it has no configured signatory`.

I recall this problem, but from the top of my head I don't remember how to approach it. I'll try to play with it at my next OSS time slot, unless you have some suggestions 🙂 We could also add extra validation to the workflow so that we fail faster if something is incorrectly set e.g. with the secrets.

**Comment by OptimumCode** [comment]
Also, for some reason, `ORG_GRADLE_PROJECT_snake-kmp.ossrhStagingRepositoryId` was set to `null` as I can see

**Comment by OptimumCode** [comment]
Regarding the problem with signing:
`> Cannot perform signing task ':signIosArm64Publication' because it has no configured signatory.`

I tried to sign the artefacts locally using my own GPG key. I had the same problem until I joined both `signing` blocks and moved them out from `afterEvaluate`. After that, the publication to the local repo was successful and all artefacts were signed.

Maybe it is no longer required to add publication signing after the script is evaluated?

**Comment by OptimumCode** [comment]
Please, ignore the above: I restored the original script state and everything works fine...

Just in case, here is how I exported and configured the signing:

1. Set up the signing key (I had to rename the property to remove dots - I had some errors with the old name)
    export ORG_GRADLE_PROJECT_signingKey=$(gpg --export-secret-keys --armor <key ID>)
    @krzema12 have you used the same command to get the key when added it to the secrets?
2. Configured the `snake-kmp.signing.keyId` to the last 8 characters of the full key ID
3. And the password of course

**Comment by OptimumCode** [comment]
@krzema12 also do you mind me re-running the staging repo job in debug mod - want to see why the staging repo ID was `null`?

**Comment by krzema12** [comment]
After fixing several secrets (thanks for detailed instructions!), we've got another error:

```
> Failed to publish publication 'iosArm64' to repository 'SonatypeRelease'
   > Could not PUT 'https://oss.sonatype.org/service/local/staging/deployByRepositoryId/null/it/krzeminski/snakeyaml-engine-kmp-iosarm64/3.0.2/snakeyaml-engine-kmp-iosarm64-3.0.2.klib'. Received status code 400 from server: Bad Request
```

Looks like we're a few steps further. The `null` in the URL indeed seems to originate from `ORG_GRADLE_PROJECT_snake-kmp.ossrhStagingRepositoryId: null`. It isn't straightforward for me to retrieve the staging profile ID, I just see a name there with the value "it.krzeminski".

> @krzema12 also do you mind me re-running the staging repo job in debug mod - want to see why the staging repo ID was `null`?

If we trust GitHub that it will mask all secrets in the logs, I guess we can try it 😊

**Comment by OptimumCode** [comment]
Looks like you need to create it first... This is a documentation I have found: https://help.sonatype.com/en/configuring-the-staging-suite.html

But looking at notice there I am not sure how up-to-date it is for your Sonstype repository. But it worth to check I think

**Comment by OptimumCode** [comment]
It seems like if you open oss.sonatype.org, go to `Staging Profiles` and click on the staging profile with the corresponding group, its ID appears in the URL in the browser (alternatively you can open dev tools in the browser and find the ID in the network requests). I have not found a better way to obtain it...

**Comment by OptimumCode** [comment]
I have also noticed that the last run was on the old version of the workflow
https://github.com/krzema12/snakeyaml-engine-kmp/actions/runs/10497100373/workflow#L92

The is not `!cancelled()` call in the condition for `drop-staging-repo` job

**Comment by OptimumCode** [comment]
Great! I am glad that I could help

**Comment by aSemy** [comment]
Really nice, thanks @OptimumCode!

**Comment by aSemy** [comment]
Now that Sonatype publishing is set up so well I think we should remove the git-branch publishing

- https://github.com/krzema12/snakeyaml-engine-kmp/blob/ef67844c272f9efe55418a3e2d5a9ed5149024d8/.github/workflows/git-branch-publish.yml
- https://github.com/krzema12/snakeyaml-engine-kmp/blob/ef67844c272f9efe55418a3e2d5a9ed5149024d8/buildSrc/src/main/kotlin/buildsrc/conventions/git-branch-publish.gradle.kts

And delete the branch https://github.com/krzema12/snakeyaml-engine-kmp/tree/artifacts


It'd probably be nice to set up an alternative, like GitHub packages, as a backup in case Maven Central suffers downtime. Despite GitHub packages' flaws, at least it's free and fairly easy to set up.
